keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,314 @@
+- name: default-token-uudge
+imagePullSecrets:
+- name: myregistrykey
+imageDecryptSecrets:
+- name: <secret name>
+- image: [<image name>]
+```
+
+For example,
+```bash
+kubectl patch serviceaccount default -p '{"imageDecryptSecrets":[{"name":"keysecret", "images":["localhost:5000/nginx:enc"]}]}'
+```
+
+#### Image Handler
+
+The privake keys are extracted from `ImageDecryptSecret` and passed to the CRI as by bundling them in `DecryptParams` as a part of `PullImageRequest`:
+
+
+```protobuf
+// DecryptParams is used to send decryption parameters to the image service of the CRI
+message DecryptParams {
+    repeated string privateKeyPasswds = 1;
+}
+```
+```protobuf
+message PullImageRequest {
+    // Spec of the image.
+    ImageSpec image = 1;
+    // Authentication configuration for pulling the image.
+    AuthConfig auth = 2;
+    // Config of the PodSandbox, which is used to pull image in PodSandbox context.
+    PodSandboxConfig sandbox_config = 3;
+    // DecryptParams for the images service of the CRI
+    DecryptParams dcparams = 4;
+}
+```
+
+See the [reference implementation](https://github.com/harche/kubernetes/blob/8a0ef9898a55110d7e41f8e1a846c66cfc2fa691/pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto#L1063)
+
+Just like in the case of `ImagePullSecrets`, CRI [forwards](https://github.com/stefanberger/containerd/blob/image-encryption.v4/vendor/github.com/containerd/cri/pkg/server/image_pull.go#L103) the private keys to containerd (or CRI-O) where the actual decryption of the image will take place. 
+
+
+### Consumption of the ImageDecryptSecrets
+
+In the diagram https://imgur.com/a/eVkUF9w, when the kubelet wants to create a new pod which has encrypted images it has to first retrieve the decryption keys referenced decryption keys (They can be referenced directly in the pod or deployment yaml or in the pod’s service account). 
+
+Kubelet sends the request to pull the image along with the decryption keys to CRI which then forwards it to containerd. Containerd looks up for the image in the snapshotter used by CRI, if the image doesn’t exist then it downloads it and decrypts it using the decryption keys that were passed by CRI. 
+
+In case containerd finds the image in the snapshotter cache which was already pulled and decrypted by earlier request, it performs a check we call ‘Image Authorization’. This check only attempts to unwrap the keys in the image manifest of the image. In simple terms it means everytime you want to use an encrypted image, you will have to prove that you have the necessary keys to decrypt it even if the actual image is present in the decrypted form in the snapshotter. This prevents an attack where a user, without having decryption keys, might get access to encrypted image content if their pod gets scheduled on a worker node where that particular encrypted image was already pulled and decrypted by earlier request. 
+
+There is no change required in CRI RuntimeService or Kubernetes
+
+## Alternatives Considered
```

Comment:
Thank you for adding the alternatives considered section.

I agree there's an advantage to doing this reusing existing patterns.